homeassistant/components/lcn/helpers.py

```diff
@@ -1,18 +1,191 @@
+    for device_config in config_entry.data[CONF_DEVICES]:
+        if device_config[CONF_UNIQUE_ID] == unique_device_id:
+            return device_config
+    return None
+
+
+def get_device_address(device_config):
+    """Return a tuple with address information."""
+    return (
+        device_config[CONF_SEGMENT_ID],
+        device_config[CONF_ADDRESS_ID],
+        device_config[CONF_IS_GROUP],
+    )
+
+
+def get_device_connection(hass, unique_device_id, config_entry):
+    """Return a lcn device_connection."""
+    device_config = get_device_config(unique_device_id, config_entry)
+    if device_config:
+        host_connection = hass.data[DOMAIN][config_entry.entry_id][CONNECTION]
+        addr = pypck.lcn_addr.LcnAddr(*get_device_address(device_config))
+        device_connection = host_connection.get_address_conn(addr)
+        return device_connection
+    return None
+
+
+def get_resource(domain_name, domain_data):
+    """Return the reosurce for the specified domain_data."""
+    if domain_name in ["switch", "light"]:
+        return f'{domain_data["output"]}'
+    if domain_name in ["binary_sensor", "sensor"]:
+        return f'{domain_data["source"]}'
+    if domain_name == "cover":
+        return f'{domain_data["motor"]}'
+    if domain_name == "climate":
+        return f'{domain_data["source"]}.{domain_data["setpoint"]}'
+    if domain_name == "scene":
+        return f'{domain_data["register"]}.{domain_data["scene"]}'
+    raise ValueError("Unknown domain.")
+
+
+def generate_unique_id(
+    address=None,
+    domain_config=None,
+):
+    """Generate a unique_id from the given parameters."""
+    unique_id = ""
+    if address:
+        is_group = "g" if address[2] else "m"
+        unique_id += f"{is_group}{address[0]:03d}{address[1]:03d}"
+        if domain_config:
+            resource = get_resource(*domain_config)
+            unique_id += f"-{resource}".lower()
+    return unique_id
+
+
+def import_lcn_config(lcn_config):
+    """Convert lcn settings from configuration.yaml to config_entries data."""
+    data = {}
+    for connection in lcn_config[CONF_CONNECTIONS]:
+        host = {
+            CONF_HOST: connection[CONF_NAME],
+            CONF_IP_ADDRESS: connection[CONF_HOST],
+            CONF_PORT: connection[CONF_PORT],
+            CONF_USERNAME: connection[CONF_USERNAME],
+            CONF_PASSWORD: connection[CONF_PASSWORD],
+            CONF_SK_NUM_TRIES: connection[CONF_SK_NUM_TRIES],
+            CONF_DIM_MODE: connection[CONF_DIM_MODE],
+            CONF_DEVICES: [],
+            CONF_ENTITIES: [],
+        }
+        data[connection[CONF_NAME]] = host
+
+    for domain, domain_config in lcn_config.items():
+        if domain == CONF_CONNECTIONS:
+            continue
+        domain_name = DOMAIN_LOOKUP[domain]
+        # loop over entities in configuration.yaml
+        for domain_data in domain_config:
+            # remove name and address from domain_data
```

Comment:
Name and address are not part of the domain specific data. So they are pulled out of the dict and set at appropriate position in config_entry.data (name in `config_entry.data[CONF_ENTITIES]`; address in `config_entry.data[CONF_DEVICES]`.

Reply:
Ah, sorry I missed that.  I see now, below where there is the `CONF_DOMAIN_DATA: domain_data.copy()`.

This block of code is a bit long given the interleaved update/creation (my other comment about using a dict instead of a list) so I think it might help to move the logic that removes them from the list in the " # create new entity_config ".  It might also even be more clear to make the copy, then remove NAME and ADDRESS?